### PR TITLE
JRMP2-2 Display list in page 

### PR DESCRIPTION
**Pull request by SvetlanaTr**
(**Done**)
- to agree on contracts (interfaces) of the data model.
- add an endpoint to the backend that will return a list of products with this interface.
- fill the test database (23 items)
- endpoint must support pagination
- the frontend must be able to receive data from this endpoint.
- draw them (design - the simplest at your discretion)
- be able to work with pagination

### CODE REVIEW

File: backend/dev/app.ts [hunk 1]
```diff
@@ -1,11 +1,13 @@
-import express from "express";
+import cors from 'cors';
+import express from 'express';
 import { router } from './router.js';
 
 const app = express();
-let port:number = process.env.PORT === undefined ? 4200 : +process.env.PORT;
+const port: number | string = process.env.PORT || 5500;
```
Comment (DmitryKorolDevPro) [inline backend/dev/app.ts]: do not branch! work in one data type.

why is "const" only now? @exORYON , this was a commentary on the previous PR

File: backend/dev/app.ts [hunk 2]
```diff
@@ -1,11 +1,13 @@
...
 
+app.use(cors())
```
Comment (DmitryKorolDevPro) [inline backend/dev/app.ts]: don't forget the semicolon

File: backend/dev/app.ts [hunk 3]
```diff
@@ -1,11 +1,13 @@
...
 app.use('/', router);
 
 app.listen(port, () => {
-  console.log(`App is working at http://localhost:${port}.`)
+  console.log(`App is working at http://localhost:${port}/api/sneakers`);
```
Comment (DmitryKorolDevPro) [inline backend/dev/app.ts]: the application works only on "http://localhost:${port}/api/sneakers" ? on "http://localhost:${port}" doesn't work?

what will you write when there are several urls?

File: package.json [hunk 4]
```diff
@@ -1,29 +1,34 @@
...
-  "name": "--",
+  "name": "e-shop",
   "version": "1.0.0",
-  "description": "",
-  "main": "index.js",
   "type": "module",
   "scripts": {
-    "backend-prod": "tsc -p backend/tsconfig.json && node backend/prod/app.js",
-    "backend-dev": "node node_modules/nodemon/bin/nodemon.js -e ts --exec \"npm run backend-prod\""
+    "start": "npm run build && start npm run backend-start && start npm run frontend-start",
```
Comment (DmitryKorolDevPro) [inline package.json]: what? think again, please)

File: package.json [hunk 5]
```diff
@@ -1,29 +1,34 @@
...
+    "build": "tsc -p frontend/tsconfig.json && tsc -p backend/tsconfig.json",
+    "backend-start": "node backend/prod/app.js",
+    "backend-dev": "node node_modules/nodemon/bin/nodemon.js -e ts --exec npm run backend-start",
+    "frontend-start": "cd frontend && live-server",
+    "frontend-dev": "tsc -p frontend/tsconfig.json --watch"
```
Comment (DmitryKorolDevPro) [inline package.json]: under what situation do you want to run "dev" and "start" separately?

File: frontend/style.css [hunk 6]
```diff
@@ -0,0 +1,147 @@
+:root {
+  --headerAndFooterHeight: 90px;
+  --black: #1c1c1c;
+  --white: #fafaff;
+  --grey: #eef0f2;
```
Comment (DmitryKorolDevPro) [inline frontend/style.css:5]: 👍

File: frontend/index.html [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+<html>
+
+<head>
+    <meta charset="utf-8" />
+      <title>SLICER e-Shop</title>
+    <meta name="keywords" content="" />
+    <meta name="description" content="" />
+    <link rel="preconnect" href="https://fonts.gstatic.com">
+    <link
+        href="https://fonts.googleapis.com/css2?family=Montserrat:ital,wght@0,100;0,200;0,300;0,400;0,600;0,700;0,800;0,900;1,400;1,500;1,900&display=swap"
```
Comment (DmitryKorolDevPro) [inline frontend/index.html]: think about optimization. do you need all these types of font? pay attention to how much the size of the downloaded font will differ

File: frontend/index.html [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+              <div class="align-center pagination__number" id="num">1</div>
+            <button class="fas fa-chevron-right pagination__button" id="next"></button>
+          </div>
+      </main>
+    <footer class="align-center footer">
+        <small>2021 SLICER ©</small>
+    </footer>
+
+    <script src="https://kit.fontawesome.com/e9e0221c2b.js" crossorigin="anonymous"></script>
+    <script src="./prod/script.js"></script>
```
Comment (DmitryKorolDevPro) [inline frontend/index.html:30]: not critical, but a point for improvement. read about "js file versioning" and why you need it.

File: backend/dev/controller.ts [hunk 9]
```diff
@@ -1,6 +1,51 @@
 import { Model } from './model.js';
 const model = new Model();
 
-console.log('Controller is working.');
+export type ResponseObj = {
+  statusCode: number,
+  result: object[]
+}
```
Comment (DmitryKorolDevPro) [inline backend/dev/controller.ts]: why "type"? interface fits more
It is better to keep interfaces in separate files (not critical)

File: backend/dev/controller.ts [hunk 10]
```diff
@@ -1,6 +1,51 @@
...
+    }
+    
+    if (response.result.length === 0) {
+      response.statusCode = 204;
+    }
+
+    return response;
+  }
+
+  async getFiveItems(page: any):Promise<ResponseObj> {
```
Comment (DmitryKorolDevPro) [inline backend/dev/controller.ts]: using the type "any" is a bad habit. you know what type you expect!

can we make a generic method for more than just 5 elements?

File: backend/dev/model.ts [hunk 11]
```diff
@@ -1,21 +1,16 @@
 import { promises as fsp } from 'fs';
 
-console.log('Model is working.');
+interface DatabaseModel {
+  getList(): object[] | object;
```
Comment (DmitryKorolDevPro) [inline backend/dev/model.ts]: method getList - can't return object, only array

File: frontend/dev/script.ts [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+    updateCatalog();
+  }
+}
+
+getDataAndUpdatePage();
+updatePaginationButtons();
+
+async function getDataFromBE(page: number) {
+  try {
+    return (await fetch(`http://localhost:5500/api/sneakers/${page}`))
```
Comment (DmitryKorolDevPro) [inline frontend/dev/script.ts:44]: domain name (localhost:5500) better save in variable. this is a value that can change

File: frontend/dev/script.ts [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+      .json()
+      .then(list => {
+        if (list.length === 0) {
+          throw new Error('Error. Stock is empty');
```
Comment (DmitryKorolDevPro) [inline frontend/dev/script.ts:48]: why is an empty list an error?

File: frontend/dev/script.ts [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+        } else {
+          return true;
+        }
+      })
+  } catch (error) {
+    return false;
+  }
+}
+
+function errorOccurred(error = ''): void {
```
Comment (DmitryKorolDevPro) [inline frontend/dev/script.ts]: function name should always be a verb. applies to all

File: frontend/dev/script.ts [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+      .then(list => {
+        if (list.length === 0) {
+          return false;
+        } else {
+          return true;
+        }
+      })
+  } catch (error) {
+    return false;
+  }
```
Comment (DmitryKorolDevPro) [inline frontend/dev/script.ts]: maybe it's worth making it easier?) if we work with pagination, the backend should immediately return data on the number of pages

File: frontend/dev/script.ts [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+    buttonNext.disabled = false;
+  } else {
+    buttonNext.disabled = true;
+  }
+
+  if (currentPage > 1) {
+    buttonPrev.disabled = false;
+  } else {
+    buttonPrev.disabled = true;
+  }
```
Comment (DmitryKorolDevPro) [inline frontend/dev/script.ts:198]: can be written shorter

buttonPrev.disabled = currentPage <= 1

File: frontend/dev/script.ts [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+}
+
+function showSpinner(): void {
+  if (contentContainer) {
+    contentContainer.insertAdjacentHTML('afterbegin', `<i class="fas fa-compact-disc spinner" id="spinner"></i>`);
+  }
+}
+
+function hideSpinner(): void {
+  const spinner: HTMLElement | null = document.querySelector('#spinner');
```
Comment (DmitryKorolDevPro) [inline frontend/dev/script.ts:130]: it is better to store all IDs in a separate object. to simplify future maintenance

File: frontend/dev/script.ts [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+  }
+}
+
+function errorOccurred(error = ''): void {
+  console.error(`Cannot load and display items. ${error}`);
+
+  if (contentContainer) {
+    contentContainer.innerHTML = '<span class="error">Sorry, try again later...</span>';
+  }
+}
```
Comment (DmitryKorolDevPro) [inline frontend/dev/script.ts:215]: the file is quite large. can we optimize its breakdown?

**Comment by DmitryKorolDevPro** [comment]
Where did the separate PR on the basis of the front disappear? why did he become a part of this story?

**Comment by DmitryKorolDevPro** [comment]
conflicts need to be fixed

**Comment by DmitryKorolDevPro** [comment]
there are still problems with commit messages. do we need to talk it over with a voice?

File: backend/dev/app.ts (hunk 1 above)
Comment (dmytro-khyzhniak) [reply backend/dev/app.ts]: "why is "const" only now?" - these changes are already merged. I don`t know why it shows like that's new fixes. Probably my bad during working with Git.

Fixed to:
`const defaultPort:number = parseInt(process.env.PORT || "");
const port:number = defaultPort || 5500;`

File: package.json (hunk 5 above)
Comment (dmytro-khyzhniak) [reply package.json]: We were using that while each of us was working alone before connecting BE and FE together.
Merged commands together

File: frontend/dev/script.ts (hunk 13 above)
Comment (SvetlanaTr) [reply frontend/dev/script.ts:48]: @DmitryKorolDevPro How to do it correctly?

**Pull request closed without merging**